oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtEncoder.java

```diff
@@ -89,28 +74,36 @@
  * @see <a target="_blank" href="https://connect2id.com/products/nimbus-jose-jwt">Nimbus
  * JOSE + JWT SDK</a>
  */
-final class NimbusJwsEncoder {
+public final class NimbusJwtEncoder implements JwtEncoder {
 
 	private static final String ENCODING_ERROR_MESSAGE_TEMPLATE = "An error occurred while attempting to encode the Jwt: %s";
 
+	private static final JwsHeader DEFAULT_JWS_HEADER = JwsHeader.with(SignatureAlgorithm.RS256).build();
+
 	private static final JWSSignerFactory JWS_SIGNER_FACTORY = new DefaultJWSSignerFactory();
 
 	private final Map<JWK, JWSSigner> jwsSigners = new ConcurrentHashMap<>();
 
 	private final JWKSource<SecurityContext> jwkSource;
 
 	/**
-	 * Constructs a {@code NimbusJwsEncoder} using the provided parameters.
+	 * Constructs a {@code NimbusJwtEncoder} using the provided parameters.
 	 * @param jwkSource the {@code com.nimbusds.jose.jwk.source.JWKSource}
 	 */
-	NimbusJwsEncoder(JWKSource<SecurityContext> jwkSource) {
+	public NimbusJwtEncoder(JWKSource<SecurityContext> jwkSource) {
```

Comment:
When support for Nested JWTs is added, you will need a constructor that takes two `JWKSource` values, one for this application's signing keys and one for the recipient's encryption keys. 

Given that, I don't think it's an issue to instead take Nimbus's `JWSMinter` as a parameter instead of the `JWKSource`. I like taking a `JWSMinter` better because it makes the intent of the constructor unambiguous, that these keys are for signing, not for encrypting.

Alternatively, you could use a static factory method, e.g. `fromSigningJwkSource(jwkSource)` or similar.

Reply:
When we add support for JWE, we can add another constructor that either takes 2x `JWKSource` or `JWSMinter` and `JWKSource`.

`JWKSource` is currently used throughout the code but there is no usage of `JWSMinter`.

Do you see any issues if we stick with `JWKSource`?

Reply:
> Do you see any issues if we stick with `JWKSource`?

One `JWKSource` seems ambiguous. If `NimbusJwtEncoder` supports encryption as well, is the single-`JWKSource` constructor for signing or for encryption?

> `JWKSource` is currently used throughout the code but there is no usage of `JWSMinter`.

I think `JWSMinter` makes more sense as a dependent component, similar to `JWTProcessor` on the `JwtDecoder` side. It would imply changing the implementation to use `JWSMinter`.

If "un-nested" JWE is not a use-case or if the ambiguity caused by a single-`JWKSource` constructor is not a concern, then the implementation could add a `JWSMinter` constructor later on, though I don't really see a reason to wait unless the Nimbus component is not yet suitable.